File: Problems/Algorithms/20. Valid Parentheses/ValidParentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()
        
        s.forEach {
            when (it) {
                '(', '[', '{' -> stack.push(it)
                else -> {
                    val end: Char = when (it) {
                        ')' -> '('
                        ']' -> '['
                        '}' -> '{'
                        else -> throw RuntimeException("Unknown entry $it")
                    }
                    
                    if (end != stack.poll()) return false
                }
            }
        }
        
        return stack.isEmpty()
    }
}

File: Problems/Algorithms/20. Valid Parentheses/valid_parentheses.py
#!/usr/bin/env python3

from collections import deque


class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()
        
        for ch in s:
            if ch in ['(', '[', '{']:
                stack.append(ch)
            else:
                end = ''
                if ch == ')':
                    end = '('
                if ch == ']':
                    end = '['
                if ch == '}':
                    end = '{'

                if len(stack) == 0:
                    return False
                if end != stack.pop():
                    return False
        
        return len(stack) == 0
